src/compiler/checker.ts

```diff
@@ -23001,6 +23005,13 @@ namespace ts {
             }
         }
 
+        function getContextualTypeForStaticPropertyDeclaration(declaration: PropertyDeclaration): Type | undefined {
+            const parent = declaration.parent.parent;
+            const parentType = isVariableDeclaration(parent) && getContextualTypeForVariableLikeDeclaration(parent);

```

Comment:
This seems very limiting. What about these cases?

- `function f(obj: Const = class { static opts = { len: 'foo' } }) { }` - class parent is a `ParameterDeclaration`
- `let { c }: { c: Const } = { c: class { static opts = { len: 'foo' } } }';` - class parent is a `PropertyAssignment`
- `let { c = class { static opts = { len: 'foo' } }: { c?: Const } = { };` - class parent is a `BindingElement`
- `let [c = class { static opts = { len: 'foo' } }]: [Const?] = [];` - class parent is a `BindingElement`

Reply:
We should be able to handle all of them with a simple `getContextualType(declaration.parent)`.

Reply:
I originally went with the smallest change, but happy to generalize...  Done.